Support mono 44.1 KHz tracks when performing  Digital Audio Extraction (DAE)

krcroft opened the pull request:
This PR wound up significantly exercising the code which lead to:
- uncovering some codec handling bugs involving small-block decodes (which I will flag to the SDL project when they resume work on SDL2-Sound)
- making our code more robust against excessive seeking (performed by `mpxplay`; a realtime CD Player that uses DAE under the hood. It's seek behavior would no-doubt send cache-less CDROM drives into a seek frenzy 〰️!).
- revealing the impact of even small changes in DAE read latency on the the DOS application (`mpxplay` in this case).
 
Addressing the above lead adding more checks (in the form of assertions), improved DEBUG messaging, and general code and type-size cleanups.

Each type of change is separated by commit; so these should be reviewed in sequence. The changes in each function are also packages as separate commits to make review easier and also as a good practice to support future git bisection. 

Regarding the DAE latency issue, some context first: the  job of our DAE function,`read(..)` is to write one frame of audio (or 2352 bytes) into the provided buffer. This frame of audio is provided via three ways:
1.  For **bin/cue**, we simply seek and read raw binary bytes into the buffer. 
1.  For **stereo-44.1** files, we have the track's codec decode those bytes directly into the buffer. 
1.  For **mono-44.1** files, we do the same, plus a tiny loop that converts from mono to stereo.

When running `mpxplay`, the first two scenarios can comfortably be managed at 36,000 DOSBox cycles.  

However, when the mono-conversion is added, `mpxplay` needs 200,000 DOSBox cycles to perform smoothly, despite the three-line-conversion loop only adding micro-seconds of latency to the `read()` step (the duration of the loop can be shown by enabling the `#define DEBUG` at the top of `cdrom_image.cpp`).  What's interesting is that you can simply comment out those three lines, and then `mpxplay` will be happy again with only 36,000 cycles - even when playing a mono track (although it will sound garbled).

Thoughts: normal CD-DA playback happens on an SDL mixer thread (compressed-file decodes are performed asynchronously with respect to the DOSBox emulation core), where as DAE reads are not threaded, and thus the DOSBox core itself is interrupted and blocked until read() returns.  Plus, mpxplay requests chunks of 24 frames at a time, so this synchronous blocking occurs for 24 calls to `read()` before the core is unblocked.

My guess here is that `mpxplay` is doing some kind of threaded playback in parallel to it performing `read()` calls. However, unbeknownst to it, the entire system is "pausing" during its read() calls, so the duration its playback thread has to run in is now massively smaller than it expects.  By cranking up the cycles, we speed up all of mpxplay's play code, eventually shrinking it runtime requirements to fit within whatever pulses of runtime we have between `read()` calls.

This is just a WAG though..  however I can't see anything wrong with the conversion from mono to stereo (famous last words?), given it's stark simplicity.

Thinking about CDROM behavior in general (beyond DAE and CD-DA): most CD drives included a read-ahead cache as well as a write-buffer (for burners). The read-ahead cache would have massively reduced the `read()` call latency.  So this might be a general optimization we perform in DOSBox: Add a `read_ahead()` function that runs on a C++11 thread whose job is to simply keep a `dequeue` list of sector objects filled (say 32 of them or roughly half-a-second worth, forward from the current `track_pos`).  This would benefit CDROM behavior in general, include binary IO performed on cooked sectors.

krcroft commented:
Suggest we complete [PR#186]( https://github.com/dreamer/dosbox-staging/pull/186) first; I will rebase this PR against that, which has a tiny bit of overlap.

krcroft commented:
Attached is a demo of the stereo vs mono DAE.  Enter the `mpxplay` directory and launch either `stereo.bat` or `mono.bat`.
[mpxplay.tar.gz](https://github.com/dreamer/dosbox-staging/files/4271253/mpxplay.tar.gz)

dreamer commented:
> Suggest we complete [PR#186](https://github.com/dreamer/dosbox-staging/pull/186) first; I will rebase this PR against that, which has a tiny bit of overlap.

I haven't started the review properly yet, but stepped through few initial commits - changes are quite large, but because they are nicely split into atomic changes, the review should go smooth.

I agree, let's finish #186 first.

krcroft commented on src/libs/decoders/mp3.cpp:
```diff
@@ -159,9 +160,9 @@ static Sint32 MP3_seek(Sound_Sample* const sample, const Uint32 ms)
-    const drmp3_uint64 frame_offset = static_cast<drmp3_uint64>(frames_per_ms) * ms;
-    const Sint32 result = drmp3_seek_to_pcm_frame(p_mp3->p_dr, frame_offset);
+    const uint64_t sample_rate = sample->actual.rate;
+    const drmp3_uint64 pcm_frame = ceil_divide(sample_rate * ms, 1000u);
```
There's a bug in the original `sample->actual.rate / 1000.0f` when dividing the `int / float`: `44100 / 1000.0f` becomes `44.0`; so seeks were slightly off when using 44.1 KHz MP3s; usually only noticeable in games that used a long single track and seeked to different segments within, such as Stellar 7. (When encoding with `lame -V7`, the sampling rate becomes 32,000 Hz and above `-V7` lame uses 24,000; so both of these were unaffected).

The integer division really tidies up the code, and lands the exact PCM frame without any decimal point or casting nonsense. (Longer term, all the audio code should use this call; but that's for a future PR).

dreamer commented on src/libs/decoders/opus.cpp:
```diff
@@ -252,7 +267,7 @@ static int32_t opus_open(Sound_Sample * sample, const char * ext)
     internal->total_time = pcm_result == OP_EINVAL ? -1 : // total milliseconds in the stream
         static_cast<int32_t>
-		(ceil_divide(static_cast<uint64_t>(pcm_result), OPUS_SAMPLE_RATE_PER_MS));
+        (ceil_divide(static_cast<uint64_t>(pcm_result), OPUS_SAMPLE_RATE_PER_MS));
```
cast-ception ;)

Perhaps `ceil_divide` should be a template function, so it could accept both unsigned and signed integers?

krcroft replied:
Actually, just looking at that code, although I tried to mirror the SDL approach, I find it quite tangled. The fact it needs all the sign casting back-and-forth indicates a code smell that we're mixing logical values (unsigned sizes) with success codes (-1's vs positive).  Going to refactor it.

krcroft replied:
Fixed in 3bbb5af3

dreamer commented on src/libs/decoders/opus.cpp:
```diff
@@ -266,39 +281,38 @@ static uint32_t opus_read(Sound_Sample * sample, void * buffer, uint32_t request
+
+    // Start the decode loop, incrementing as we go
+    while(remaining_samples > 0) {
+        const int result = op_read(of, buf_pos, remaining_samples, nullptr);
```
Hey, that's the suggestion I made earlier in this review :) Nice!

krcroft replied:
:+1:

krcroft commented on src/libs/decoders/mp3_seek_table.cpp:
```diff
@@ -99,13 +100,13 @@ using std::ofstream;
     struct stat stat_buf;
     int rc = stat(filename, &stat_buf);
-    return rc == 0 ? stat_buf.st_size : -1;
+    return rc == 0 ? static_cast<uint64_t>(stat_buf.st_size) : 0u;
```
Reverted back to returning -1, but now using the correct return type and avoid more casting. 
Commit 28fc76a3

krcroft commented on src/libs/decoders/opus.cpp:
```diff
@@ -195,106 +187,140 @@ static __inline__ void output_opus_info(const OggOpusFile* of, const OpusHead* o
+            sample_buffer + decoded_samples,     // buffer offset to save decoded samples
+            requested_samples - decoded_samples, // remaining samples to decode
+            nullptr);                               // link index, which we don't use
+        // Check the result code
```
(All addressed in a later commit, which you mentioned)

dreamer commented on src/dos/cdrom_image.cpp:
```diff
@@ -788,19 +788,38 @@ bool CDROM_Interface_Image::ReadSectors(PhysPt buffer,
 
+	// Resize our underlying vector if it's not big enough
+	if (readBuffer.size() < requested_bytes)
+		readBuffer.resize(requested_bytes);
```
Ah, scratch this comment; `readBuffer` is used only in this method, nowhere else, so this is all ok.

krcroft commented on src/dos/cdrom_image.cpp:
```diff
@@ -788,19 +788,38 @@ bool CDROM_Interface_Image::ReadSectors(PhysPt buffer,
+	uint8_t* buffer_position = readBuffer.data();
+
+	// Read until we have enough or fail
+	while(bytes_read < requested_bytes) {
```
argh; thanks!

dreamer replied:
I started initial work on clang formatter rules, so we'll have a mechanism to detect formatting "problems" like this before review :)

krcroft commented on src/dos/cdrom_image.cpp:
```diff
@@ -249,51 +249,102 @@ bool CDROM_Interface_Image::AudioFile::read(uint8_t *buffer,
+		const int32_t elapsed_us = static_cast<int32_t>
+		    (duration_cast<microseconds>(end - begin).count());
+		LOG_MSG("CDROM: converted %u mono to %u samples in %d us",
+	             mono_samples, mono_samples * REDBOOK_CHANNELS, elapsed_us);
```
Interesting; I suppose all of the fixed-size types map to an equivalent PR* print string.  Yes; I agree.. and the fact that neither gcc or Clang warn about this "future size change risk", even at their maximum warning sensitivity, says that that using `%d / %u` and `%ld / %lu` are essentially as reliable as the language specification.

dreamer replied:
Compilers can't warn about it, because PRI* macros are expanded by preprocessor - they literally don't see the difference - at least as long as all compilers use the same size for int (which they do, AFAIK - but language specification does not require int to be 32-bit). So `%d` for `int32_t` will work. But `%ld` for `int64_t` won't - because at least on Windows `long` is the same size as `int` :/ (that's why I replaced `Bitu` with `unsigned` in one of recent rounds of fixes for formatting warnings).

In this specific case I think maybe `int` could've been used instead of `int32_t`, but it's a non-issue to me - we can leave it as it is.

krcroft commented on src/dos/cdrom.h:
```diff
@@ -100,11 +100,11 @@ class CDROM_Interface
 	virtual bool GetAudioStatus     (bool& playing, bool& pause) = 0;
 	virtual bool GetMediaTrayStatus (bool& mediaPresent, bool& mediaChanged, bool& trayOpen) = 0;
-	virtual bool PlayAudioSector    (uint64_t start, uint64_t len) = 0;
+	virtual bool PlayAudioSector    (const uint32_t start, uint32_t len) = 0;
```
Let's look at it relative to the function body: `const` applied to **any** function argument has consistent behavior, and documents the fact that the inbound variable shall have it's value locked read-only for the entire duration of the function.  New maintainers can immediately trust that fact, even big nasty winding functions.  

So because it's behavior is consistent (even for scalars), it has value right there - it  documents the use of the variable through the life-time of the function.

When I come back to my own code (often after some time and I'm practically new to it again), I immediately know that my non-const arguments are being changed in the code, perhaps as a decrementing counter or for some dual-purpose.

In my mind, this is all separate from whatever side-effects the function can impart outside it's scope via pointer arguments and references.  

In perfect code, you've got read-only variables inboud to the function, and the new result(s) outbound from the function. The language under the hood should take care of efficiently minimizing copies-of-new-variables and so on, so it behaves.   

``` c++
#include <iostream>

using namespace std;

int change_test(const int val) {
    val++;
    return val;
}

int main()
{
    cout << change_test(1);
    return 0;
}
```

Returns:

```
main.cpp: In function ‘int test(int)’:
main.cpp:14:9: error: assignment of read-only parameter ‘d’
```

dreamer replied:
Overall I am a big fan of using const whenever possible - I just find it a bit weird when used for non-pointer, non-reference function arguments in C/C++ - so it's very specific user complaint ;)

It was only a style suggestion, so we can close this topic and leave the code as it is now :)

krcroft commented:
> let's move is_signed/unsigned checks to static asserts? This way instead of error "template argument deduction/substitution failed: ", we can have nice errors like: "error: static assertion failed: Second parameter should be unsigned"

@dreamer - those `constexpr`'s are a thing of beauty! Evaluated at compile time, more intuitive messaging, and the function separation still guarantees we will never mix differently-signed types. These two should be the new stack-exchange winners for C++11 :-)

dreamer commented:
> > let's move is_signed/unsigned checks to static asserts? This way instead of error "template argument deduction/substitution failed: ", we can have nice errors like: "error: static assertion failed: Second parameter should be unsigned"
> 
> @dreamer - those `constexpr`'s are a thing of beauty! Evaluated at compile time, more intuitive messaging, and the function separation still guarantees we will never mix differently-signed types. These two should be the new stack-exchange winners for C++11 :-)

`constexpr` is probably the best feature in C++11 :) in C++14 it got expanded a bit more, by allowing more complex constexpr functions - that would allow moving some chunks of code from runtime to compile-time (I've seen at least two implementation areas, that would significantly benefit from this - probably there's more).

krcroft commented:
No more adjustments from my side!  Last 4 commits addressed comments, so could probably be squahed; but that's cosmetic. You're call!  Thanks @dreamer!

dreamer commented:
Merged-in - thanks! :)
